src/DependencyInjection/ProophEventStoreExtension.php

```diff
@@ -39,104 +39,88 @@ public function getConfiguration(array $config, ContainerBuilder $container)
-                'prooph_event_store.projection_managers_locator',
-                new Definition(ServiceLocator::class, [$projectionManagersLocator])
-            )
-            ->addTag('container.service_locator');
-
-        $container
-            ->setDefinition(
-                'prooph_event_store.projection_manager_for_projections_locator',
-                new Definition(ServiceLocator::class, [$projectionManagerForProjectionsLocator])
-            )
-            ->addTag('container.service_locator');
+        self::defineServiceLocator($container, 'prooph_event_store.projection_managers_locator', $projectionManagersLocator);
+        self::defineServiceLocator($container, 'prooph_event_store.projection_manager_for_projections_locator', $projectionManagerForProjectionsLocator);
+        self::defineServiceLocator($container, 'prooph_event_store.projection_read_models_locator', $readModelsLocator);
+        self::defineServiceLocator($container, 'prooph_event_store.projections_locator', $projectionsLocator);
+    }
 
-        $container
-            ->setDefinition(
-                'prooph_event_store.projections_locator',
-                new Definition(ServiceLocator::class, [$projectionsLocator])
-            )
-            ->addTag('container.service_locator');
+    private static function defineProjectionManager(ContainerBuilder $container, string $serviceId, array $config): void
+    {
+        $projectionManagerDefinition = new ChildDefinition('prooph_event_store.projection_definition');
+        $projectionManagerDefinition
+            ->setFactory([new Reference('prooph_event_store.projection_factory'), 'createProjectionManager'])
+            ->setArguments([
+                new Reference($config['event_store']),
+                isset($config['connection']) ? new Reference($config['connection']) : null,
+                $config['event_streams_table'],
+                $config['projections_table'],
+            ]);
+
+        $container->setDefinition($serviceId, $projectionManagerDefinition);
     }
 
-    public function loadProjections(array $config, string $projectionManager, ContainerBuilder $container)
+    private static function defineServiceLocator(ContainerBuilder $container, string $id, array $serviceMap): void
     {
-        foreach ($config['projections'] as $projectionName => $projectionConfig) {
-            $tag = [
-                'projection_name' => $projectionName,
-                'projection_manager' => $projectionManager,
-            ];
+        $definition = new Definition(ServiceLocator::class, [$serviceMap]);
+        $definition->addTag('container.service_locator');
+        $container->setDefinition($id, $definition);
+    }
 
+    private static function collectProjectionsForLocators(
+        array $projections,
+        string $projectionManagerId,
+        array &$projectionManagerForProjectionsLocator,
+        array &$projectionsLocator,
+        array &$readModelsLocator
```

Comment:
I'd get rid of the references in favor of this:

```
return [$projectionManagerForProjectionsLocator, $projectionsLocator, $readModelsLocator];

[$projectionManagerForProjectionsLocator, $projectionsLocator, $readModelsLocator] = self::collectProjectionsForLocators($projectionManagerConfig['projections'], $projectionManagerId,);
```

Reply:
@gquemener your thoughts about out-parameters vs array-return?